Add Realtime Sanitizer (RTsan) support

ahoarau opened the pull request:
## Description

This PR adds initial support for the `Realtime Sanitizer (RTSan)` that appeared in clang 20. It allow to check for the "realtime-safeness", by detecting dynamic memory allocations and source of non-determinism (like mutex lock, systems calls etc).
Before going all-in and instrument all the code with `[[clang::nonblocking]]`, we start by writing a unittest that check for a the main  API functions. 

To test:
```
pixi run -e sanitizers test
```

Typical output might look like this:
<img width="2310" height="725" alt="image" src="https://github.com/user-attachments/assets/db2a0964-1e3a-405e-bff5-d56b9b624fcd" />


https://clang.llvm.org/docs/RealtimeSanitizer.html
https://clang.llvm.org/docs/FunctionEffectAnalysis.html
https://github.com/realtime-sanitizer/rtsan

## Checklist

- [x] I have run `pre-commit run --all-files` or `pixi run lint`
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary
- [x] I have made corresponding changes to the doxygen documentation
- [x] I have added tests that prove my fix or feature works
- [x] I have updated the [CHANGELOG](https://github.com/stack-of-tasks/pinocchio/blob/devel/CHANGELOG.md) or added the "no changelog" label if it's a CI-related issue
- [x] I have updated the [README credits section](https://github.com/stack-of-tasks/pinocchio?tab=readme-ov-file#credits)

ahoarau committed:
add sanitizers environnement to enable RTsan

dont build benchmarks

make build depends on configure

ahoarau committed:
update pixi.lock

ahoarau committed:
add BUILD_WITH_RTSAN option to enable the unittest
later this could be used to instrument the whole library

format cmake

rewrite rtsan cmake comments

ahoarau committed:
add test for dynamic allocations with rtsan

format dynamic-allocations.cpp

add more models to test
test is now failing in multiple parts

refactor rtsan test suite to only allow known
allocation-free functions
Joint Composites and Mimic causes dynamic
allocations for jacobian fonction
(and all the others).
Only the Fk is fully safe so far.

only enable nonallocating
static variables, depending on the variable type,
can be implemented with an underlying
pthread_mutex_lock.
This is the case on mac-arm64, where static int
is most probably implemented with an atomic,
but static MyClass is protected by a mutex,
causing rtsan to trigger an with nonblocking.
nonallocating is less strict, but still useful.

ahoarau committed:
ignore .ruff_cache/

format .gitignore

ahoarau committed:
add Realtime Sanitizer in CI

only run on ubuntu-latest

add fail-fast strategy for rtsan build

continue on error (rtsan)

add rtsan build in the check list

display correct name on RTsan CI

only use run test

use || true as continue on error is not working as expected

build first, then test

display a warning when rtsan fails

using matrix.name in RTSan build

enable rtsan in CI by default
now that rtsan tests handle joint composites and mimics properly, the CI can be enabled by default

ahoarau committed:
update changelog

jorisv commented on unittest/dynamic-allocations.cpp:
```diff
@@ -0,0 +1,468 @@
+
+  // Build model
+  Model model;
+  buildModels::humanoidRandom(model, true);
```
We should parametrize this test with different model.
Joint (mimic, composite) can have some hidden allocation.
Also, mimic only support the following algorithms:
- RNEA
- CRBA
- Forward Kinematics
- Jacobians and Frames
- Centroidal Algorithm (ccrba)

ahoarau replied:
Added tests with additional robots. Allocations detected !

jorisv replied:
Mimic are only supported in few algorithm (listed in my previous comment).
Unsupported algo will throw and then biasing the test.

ahoarau replied:
refactored the test to only execute certain tests if the model has composites or mimic joints. 👍 
I'm waiting for the green CI, then I'll squash some commits

jorisv commented on pixi.toml:
```diff
@@ -349,3 +349,61 @@ new-version = { features = [
+  "-DCMAKE_BUILD_TYPE=Debug",
+  "-DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX",
+  "-DBUILD_TESTING=ON",
+  "-DBUILD_BENCHMARK=ON",
```
Why did you build benchmark ?

ahoarau replied:
turned to OFF 👍

jorisv commented on .github/workflows/macos-linux-windows-pixi.yml:
You can add you job in this list (and pinocchio-pixi-build and   pinocchio-python-standalone-pixi that I have forget)

ahoarau replied:
done 👍

jorisv commented on .github/workflows/macos-linux-windows-pixi.yml:
```diff
@@ -278,3 +278,20 @@ jobs:
+  rtsan-build:
+    name: Realtime Sanitizer
+    runs-on: [ubuntu-latest, macos-latest]
+
```
```suggestion

    strategy:
      fail-fast: false
      
```

ahoarau commented:
Added a warning annotation in the CI when RTSan unittest failed.

Squashed the commits.

<img width="430" height="183" alt="warn" src="https://github.com/user-attachments/assets/14d878d0-0c2f-4bbb-9f68-c62f4901aae6" />

jorisv commented on .github/workflows/macos-linux-windows-pixi.yml:
```diff
@@ -264,12 +263,51 @@ jobs:
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - os: ubuntu-latest
+            name: Ubuntu (x86_64)
+
+          - os: macos-latest
+            name: MacOS (ARM64)
```
```suggestion
        os: [ubuntu-latest, macos-latest]
```

`name` is never used.

ahoarau replied:
Using name instead 👍

ahoarau committed:
enable joint torque regressors and contact dynamics tests for robots wo micmi/composite joints
